### DISCUSSION TITLE
gurobi solution modeling

### DISCUSSION THREAD

**Discussion by wnn666666** [yalmip/YALMIP#1374] [Optimization modelling]
The solution found by using gurobi+ is not optimal, the output of decision variables is all upper limit, the yalmip version is 2021.03.31, and the gurobi version is 10.0.3. My problem is the same as the problem discussed before when the time is 2023,04,21. After turning off presolve, the result remains unchanged. How to solve this problem, and the installation of yalmip version and gurobi version has anything to do with it?

**Comment by johanlofberg** [comment]
So how do you know the solution isn't optimal?

And what do you mean with gurobi+. If you selected 'gurobi' as solver, and yalmip told you it cannot use 'gurobi', and you then selected 'gurobi+', it just means you asked yalmip to create a nonsense model for gurobi, and thus a nonsense solution will be returned

**Comment by wnn666666** [comment]


**Reply by johanlofberg** [reply]
gurobi cannot deal with that, it is not SOCP-representble (unless x or y is integer in which case it can be massaged into something MILP-representable)

**Comment by wnn666666** [comment]


**Reply by johanlofberg** [reply]
If x or y is integer it can be rewritten using MILP. If not, it  cannot be solved using Gurobi unless approximated

NEVER use the + operator to select solver. It means "Hey YALMIP, you are incorrectly claiming that gurobi is not capable of solving this problem. I know better than you so I force you to setup a model which you claim gurobi cannot handle, and send that model to gurobi, and then give me back what ever it returns, despite it telling me that it cannot solve the problem"

**Comment by wnn666666** [comment]


**Reply by johanlofberg** [reply]
MILP is mixed-integer linear programming. If x or y are binary or integer, you simply model it using implications

https://yalmip.github.io/tutorial/logicprogramming#:~:text=y%20%3D%20f(x)%2C%20x%20scalar%20integer

Basically "if x = 0 then obj = y, if x=1 then obj = 2*y" etc

**Comment by wnn666666** [comment]


**Reply by johanlofberg** [reply]
Integers are just more implications. 2^x*y is simply "if x = 0 then obj = y, if x = 1, then obj = 2*y, if x = 3 then obj = 8*y" etc

**Comment by wnn666666** [comment]


**Reply by johanlofberg** [reply]
Yes, using implies and possibly new variables to massage things to something MILP (or MIQP)-representable

**Comment by wnn666666** [comment]


**Reply by johanlofberg** [reply]
I have no idea what you are talking about

**Comment by wnn666666** [comment]


**Reply by johanlofberg** [reply]
You will have to come up with a way to model your situation without using variable whose structure depends on another variable. It is simply not possible to model it that way. Typically, you will have to use a variable whose length is the longest possible, and then model using that, and impose stuff to represent the stuff that you really want to happen when the length is only k (forcing remaining variables to be zero etc)

**Comment by wnn666666** [comment]


**Reply by johanlofberg** [reply]
Unclear what you mean, but maybe this https://yalmip.github.io/tutorial/logicprogramming#:~:text=much%20as%20possible-,Counters%20and%20loops,-A%20very%20common

**Comment by wnn666666** [comment]


**Reply by johanlofberg** [reply]
I have no idea what you are asking about

**Comment by wnn666666** [comment]


**Reply by johanlofberg** [reply]
you're not showing the error

**Comment by wnn666666** [comment]


**Reply by johanlofberg** [reply]
exponentials have to be approximated https://yalmip.github.io/command/interp1/#:~:text=Example%20(Linear%20optimization)

**Comment by wnn666666** [comment]


**Reply by johanlofberg** [reply]
You cannot use decision variables in programming constructs (for-loops etc). Everything has to be written using linear operations or high-level logic conditions implemented using implies (i.e. mixed-integer linear operations)

**Comment by wnn666666** [comment]


**Reply by johanlofberg** [reply]
what would break even mean without there being a for (or while) loop somewhere. break what?

**Comment by wnn666666** [comment]
When calling gurobi in yalmip for solution, the objective function contains (2^x)*y, and x and y here are intvar sdpvar variables respectively. If you do not use implication, how to deal with approximation, is there any example

**Reply by johanlofberg** [reply]
since you're already in the integer programming world, I don't see why you would want to deal with this using some crappy approximation, as it is trivial to represent exactly

**Comment by wnn666666** [comment]
Thanks for your reply, because the data of integer x at this time is a little huge, if you use implication to find out the minimum need to simulate several times, so I want to approximate an optimal solution

**Reply by johanlofberg** [reply]
no idea what you mean with simulate several times.

a simple binary model would be the easiest. if you maximize (2^x)y it can be represented with a complicated geomean model. I don't see an approximation being performant

**Comment by wnn666666** [comment]
implies is used in the model, and Gurobi is chosen to solve it. From the output lp file, implies that the default M value is 10000, is there a way to correct this value

**Reply by johanlofberg** [reply]
If YALMIP has resorted to the bound 10^4 it means you are not bounding variables sufficiently well or you have bad scaling in your model. If you against my advice would want to force a larger M and thus create even more numerically unstable and weak relaxations you will have to edit derivebounds.m

**Comment by wnn666666** [comment]
Thank you, Sir. I opened the derivebounds.m file. Do I change the 1e4 Settings here.----------
M(isnan(M)) = inf;
M(isinf(M)) = 1e4; M = M+b;%+0.01;
m(isnan(m)) = -inf;
m(isinf(m)) = -1e4;m = m+b;%-0.01;

**Reply by johanlofberg** [reply]
yes

But as I said, you should improve your model so sensible bounds can be derived and not force a specific M

**Comment by wnn666666** [comment]
Hello teacher, I have two models, one is a general model, whose optimization variables are represented by intvar, for verification; the other is a deterministic model, whose corresponding variables are input by the optimization result value of the first general model intvar variable, all of which call gurobi, and the solver parameters and other information of both are consistent. The difference is only that the value of a variable is intvar, and the value of a variable is input with the value of the optimization. The optimal value of the output of the two variables is all the same, but there are differences in some decision variables, which will not affect the target result. Can we consider these two models consistent, or what causes them

**Reply by johanlofberg** [reply]
aren't you simply saying you have a problem with a non-unique optimal solution

**Comment by wnn666666** [comment]
Thank you for your reply, do you mean the case of multiple solutions? It is indeed a bit like the case of multiple solutions. The difference between the two models is only that, in order to verify the general model, the output result of the general model optimization variable intvar is directly input to the deterministic model as the value double. The rest is the same, the objective function of both is the same. There are small differences in the remaining decision variables that run but have no effect on the objective function, or how to determine that this difference is due to multiple solutions rather than differences in optimization using intvar and double modeling respectively.

**Reply by johanlofberg** [reply]
no one will know the answer to this better than you. you use two different models, and get the same solution on the important part but different on other. if you think that is fine, well that is up for you to decide

**Comment by wnn666666** [comment]
Hello,&nbsp;the&nbsp;default&nbsp;implies&nbsp;in&nbsp;yalmip&nbsp;that&nbsp;the&nbsp;large&nbsp;M&nbsp;is&nbsp;10000,&nbsp;and&nbsp;what&nbsp;about&nbsp;the&nbsp;variable,&nbsp;is&nbsp;there&nbsp;a&nbsp;limit&nbsp;on&nbsp;optimizing&nbsp;the&nbsp;variable








------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "wnn2201"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2024年4月13日(星期六) 下午3:44
***@***.***&gt;;

主题:&nbsp;回复： [yalmip/YALMIP] gurobi solution modeling (Discussion #1374)



Thank you, Sir. I opened the derivebounds.m file. Do I change the 1e4 Settings here.----------
M(isnan(M)) = inf;
M(isinf(M)) = 1e4; M = M+b;%+0.01;
m(isnan(m)) = -inf;
m(isinf(m)) = -1e4;m = m+b;%-0.01;

**Reply by johanlofberg** [reply]
?

**Comment by wnn666666** [comment]
Teacher if the variables don't currently have upper and lower bounds set on them, then yalmip's default bounds on variables are 1e4 in derivebounds, which means the maximum value of the variable can only be taken to 1e4 here


wn
***@***.***







&nbsp;




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "yalmip/YALMIP"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2024年4月13日(星期六) 下午3:35
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [yalmip/YALMIP] gurobi solution modeling (Discussion #1374)





 
If YALMIP has resorted to the bound 10^4 it means you are not bounding variables sufficiently well or you have bad scaling in your model. If you against my advice would want to force a larger M and thus create even more numerically unstable and weak relaxations you will have to edit derivebounds.m

**Comment by wnn666666** [comment]
Teacher if the variables don't currently have upper and lower bounds set on them, then yalmip's default bounds on variables are 1e4 in derivebounds, which means the maximum value of the variable can only be taken to 1e4 here，right?


wn
***@***.***







&nbsp;




------------------ 原始邮件 ------------------
发件人:                                                                                                                        "yalmip/YALMIP"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2024年4月13日(星期六) 下午3:35
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [yalmip/YALMIP] gurobi solution modeling (Discussion #1374)





 
If YALMIP has resorted to the bound 10^4 it means you are not bounding variables sufficiently well or you have bad scaling in your model. If you against my advice would want to force a larger M and thus create even more numerically unstable and weak relaxations you will have to edit derivebounds.m

**Reply by johanlofberg** [reply]
Yes, and if you cannot improve your model you will have to hard-code a change in derivebounds.m
